# Import Libraries

In [55]:
import pandas as pd
import numpy as np
import os

# Import Data

In [56]:
# Folder path
path = '/Users/katherinevaldivia/Desktop/Career Foundry Data Analytics Course work/Data Analytics Immersion/Achievement 4/06-23-Instacart Basket Analysis /'
path

'/Users/katherinevaldivia/Desktop/Career Foundry Data Analytics Course work/Data Analytics Immersion/Achievement 4/06-23-Instacart Basket Analysis /'

In [57]:
# Import products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [58]:
# Import orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Consistency Check

## Mixed-Type Data

In [59]:
# Create a new test dataframe
# The first command, df_test = pd.DataFrame(), creates a new dataframe called df_test. 

df_test = pd.DataFrame()

In [60]:
# Create a mixed type column
# The second command, df_test['mix'] = ['a', 'b', 1, True], creates a new column, mix, within df_test and fills it with numeric, string, and boolean values.

df_test['mix'] = ['a', 'b', 1, True]

In [61]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [62]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [63]:
# Changing data type to string

df_test['mix'] = df_test['mix'].astype('str')

## Missing Values

In [64]:
# This assigns the function isnull() to the df_prods dataframe, then sums the result with the attached sum() function.
# The isnull() function is used to find missing "observations", referring to entries in your dataframe.
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [65]:
# To actually view these 16 missing values, you can create a subset of the dataframe.

df_nan = df_prods[df_prods["product_name"].isnull() == True]

df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

### Observation: Records with missing product names will be filtered out and a new dataframe created.

In [66]:
#  The shape function returns the number of rows and columns in a dataframe.

df_prods.shape

(49693, 5)

In [67]:
# Create dataframe without missing values in 'product_name' column
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [68]:
# New dataframe should have exactly 16 less rows than the original dataframe.
df_prods_clean.shape

(49677, 5)

In [69]:
# Check again for mixed-type data after filtering out missing values
for col in df_prods_clean.columns.tolist():
  weird = (df_prods_clean[[col]].applymap(type) != df_prods_clean[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods_clean[weird]) > 0:
    print (col)

## Duplicates

### Finding Duplicates

In [70]:
# The code BELOW creates a new subset of df_prods_clean: df_dups, containing only rows that are duplicates.
# The duplicated() function is what identifies duplicate rows.
# It’s run on the df_prods_clean dataframe. Any duplicate rows that it finds are saved within the new df_dups dataframe.

In [71]:
# Finding full duplicates within dataframe:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [72]:
# View duplicates subset
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

### Addressing Duplicates

In [73]:
# Now that you’ve identified the duplicate rows within your dataframe, you need to delete them. Use this handy function: 
#df.drop_duplicates()

In [74]:
# Before doing so, let’s check the current number of rows in your df_prods_clean dataframe.
#So that you can compare the number after removing the duplicates.

df_prods_clean.shape

(49677, 5)

In [75]:
# Creating a new dataframe with duplicates removed
# df_prods_clean_no_dups that contains only the unique rows from df_prods_clean.

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [76]:
# The five duplicates have been successfully deleted!

df_prods_clean_no_dups.shape

(49672, 5)

# Exercise 4.5: Consistency Checks Task

### Step 1

In [77]:
# Perform the consistency checks covered in this Exercise on your df_prods dataframe. 

### Step 2

In [78]:
# Run the df.describe() function on your df_ords dataframe. 
#share in a markdown cell whether anything about the data looks off or should be investigated further.
#Tip: Keep an eye on min and max values!

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_last_order  
count       3.421083e+06       3.421083e+06           3.214874e+06  
mean        2.776219e+00       1.345202e+01           1.111484e+01  
std         2.046829e+00       4.226088e+00           9.206737e+00  
min         0.000000e+00       0.000000e+00           0.000000e+00  
25%         1.000000e+00       1.000000e+01           4.000000e+00  
50%         3.000000e+00       1.300000e+01           7.000000e+00  
75%         5.000000e+00       1.600000e+01           1.500000e+01  
max         6.000000e+00       2.300000e+01           3.000000e+01

In [79]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### The max value for the 'prices'  column seems high and should be investigated further.

In [80]:
# Investigate record with max 'prices' value
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices'] == 99999.000000]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

### Step 3

In [81]:
# Check for mixed-type data in your df_ords dataframe.

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col, ' mixed')
  else: print (col, ' consistent')

Unnamed: 0  consistent
order_id  consistent
user_id  consistent
order_number  consistent
order_day_of_week  consistent
order_hour_of_day  consistent
days_since_last_order  consistent


#### No mixed type data has been detected for df_ords

### Step 4

In [82]:
# If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

#### No mixed-type data was found.

### Step 5

In [83]:
# Run a check for missing values in your df_ords dataframe.
# In a markdown cell, report your findings and propose an explanation for any missing values you find.
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
order_day_of_week             0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In [84]:
# Create subset to show missing values in df_ords
df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [85]:
# View missing values subset
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
11               11   2168274        2             1                  2   
26               26   1374495        3             1                  1   
39               39   3343014        4             1                  6   
45               45   2717275        5             1                  3   
...             ...       ...      ...           ...                ...   
3420930     3420930    969311   206205             1                  4   
3420934     3420934   3189322   206206             1                  3   
3421002     3421002   2166133   206207             1                  6   
3421019     3421019   2227043   206208             1                  1   
3421069     3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

#### The records containing missing values in 'days_since_last_order' column are also the ones with a value of 1 in the 'order_number' column. This makes sense since because this is the customers first order so they will have a missing value in the 'days_since_last_order' column. Also, the maximum user id is 206209 and total missing values for days_since_prior_order is also 206209, meaning each user will have one NaN value located in their first order.

### Step 6

In [86]:
# Address the missing values using an appropriate method.
#In a markdown cell, explain why you used your method of choice.


#### I decided no to make any changes because those missing values are useful in determining that those customers are new customers.

### Step 7

In [87]:
# Run a check for duplicate values in your df_ords data.
# In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

df_ords_dups = df_ords[df_ords.duplicated()]

In [88]:
# View df_ords_dups 

df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

#### No Duplicates were found.

### Step 8

In [89]:
# Address the duplicates using an appropriate method.

####  No duplicates were found so there is nothing to address.

### Step 9 Export Dataframes

In [90]:
# Export df_prods_clean_no_dups

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [91]:
# Export final df_ords 

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))